### PR TITLE
Core policy specification

### PR DESCRIPTION
**Pull request by vdeturckheim**
Here is a first draft of specification for core policy.

See #327
cc @drifkin @mhdawson @bmeck @mikesamuel @deian

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -126,4 +126,20 @@ The [Node.js Code of Conduct](https://github.com/nodejs/admin/blob/master/CODE_O
 
 The [Node.js Moderation Policy](https://github.com/nodejs/admin/blob/master/Moderation-Policy.md) applies to this WG.
 
+# Code of Conduct
```
Comment (lirantal) [inline README.md]: these are duplicates of existing code of conduct and moderation policies already in the document

File: design/core_policies.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Node.js core Security Policies
```
Comment (lirantal) [inline design/core_policies.md]: why is the directory called design ?

File: README.md (hunk 1 above)
Comment (vdeturckheim) [reply README.md]: Git sticky fingers on my side, will remove

File: design/core_policies.md (hunk 2 above)
Comment (vdeturckheim) [reply design/core_policies.md]: So, I am not sure this should go in "process" but so far I had not better idea

**Comment by drifkin** [comment]
Thanks for writing this up! I do think that we still have unanswered questions about the attacker model: are we trying to disable access to particular resources for well meaning, yet buggy code, or are we trying to disable access to resources so that malicious code cannot possibly have access to those resources?

To illustrate why this is tricky, consider disabling filesystem access. Even if you disable filesystem access for user code, Node still needs to load the user's code in the first place! Also, you may still want `require` to work for user code, in which case you're providing limited filesystem access. Node in its current state doesn't provide enough isolation to safely allow some code to have access and other code to not. You'd either have to "fix" (I'm using scare quotes because it's not really broken, it's just that Node core has always considered user code to be extremely privileged: you can run arbitrary commands using child process, after all) lots of Node code (for example, @deian is giving a talk at Node summit about crafting attacks using the binding layer, which is currently considered to be a private API) and be extremely strict on every code review; or you can put a secure, isolated layer on top like what we're doing at Intrinsic, but that's a ton of work.

**Comment by vdeturckheim** [comment]
@drifkin right that's all questions we need to answer. I mostly opened this PR so we have a brainstorm ready for Thursday afternoon.

My canonical example is "how could have we prevented the ESLint situation". Disabling network access would have done the trick here.

**Comment by drifkin** [comment]
Yeah, for sure, we’ll definitely talk a lot about these issues.

Re: the eslint issue, remember that it was an install time script that npm ran, which can be an arbitrary (even non-Node.js) command. These policies wouldn’t help against a malicious user trying to exfiltrate data (they could, for example, just use curl or manually run Node with network access enabled).

File: design/core_policies.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+Question:
+* Do we want to allow policy definition from a file
+* How can we make sure modules like linters, that people start from `.bin` never have access to network
+
+
+## Pre-built-policies
+
+| flag | impact |
+|------|--------|
```
Comment (vdeturckheim) [inline design/core_policies.md]: @cjihrig suggested preventing bindings altogether

File: design/core_policies.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+Also, a per core module blacklist can be set up:
+
+```
+$ node --disable=net,http,https,http2,dns,udp index.js
+```
+
+Question:
+* Do we want to allow policy definition from a file
+* How can we make sure modules like linters, that people start from `.bin` never have access to network
```
Comment (bmeck) [inline design/core_policies.md]: This should discuss if the scope of permissions is on files within `node_modules/.bin` or if it is in `node_modules/linter/*`. In particular I have concerns about if the permissions are self defined within `node_modules/linter` in a way that allows `linter` to alter the permissions after the first run. I think establishing the authority structure is potentially going to need more care than the feature flags.

File: design/core_policies.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+```
+
+Also, a per core module blacklist can be set up:
+
+```
+$ node --disable=net,http,https,http2,dns,udp index.js
+```
+
+Question:
+* Do we want to allow policy definition from a file
```
Comment (bmeck) [inline design/core_policies.md]: are we talking about the:

* requested policies for a given package being put into a file
* the flags enabling features to be put into a file
* the authorized pre-authorized permissions for features being placed into a file

It seems like we might need more than 1 file and/or need to clarify which parts need to live where

**Pull request closed without merging**